File: app/src/main/java/model/Todo.java
package model;

import androidx.annotation.Keep;
import androidx.room.ColumnInfo;
import androidx.room.Entity;
import androidx.room.Ignore;
import androidx.room.PrimaryKey;

import java.io.Serializable;
import java.util.Objects;

import ir.hamsaa.persiandatepicker.api.PersianPickerDate;
import ir.hamsaa.persiandatepicker.date.PersianDateImpl;
import utils.DateHelper;

@Keep
@Entity(tableName = "todos")
public class Todo implements Comparable<Todo>, Serializable {

    @ColumnInfo(name = "id")
    @PrimaryKey(autoGenerate = true)
    private int id;

    @ColumnInfo(name = "title")
    private String title;

    @ColumnInfo(name = "category_id")
    private int categoryId;

    @ColumnInfo(name = "category")
    private String category;

    @ColumnInfo(name = "priority")
    private Priority priority;

    @ColumnInfo(name = "is_done")
    private boolean isDone;

    @ColumnInfo(name = "arrive_date", defaultValue = "0")
    private long arriveDate;

    @Ignore
    private DateTime dateTime;

    public int getId() {
        return id;
    }

    public void setId(int id) {
        this.id = id;
    }

    public String getTitle() {
        return title;
    }

    public void setTitle(String title) {
        this.title = title;
    }

    public int getCategoryId() {
        return categoryId;
    }

    public void setCategoryId(int categoryId) {
        this.categoryId = categoryId;
    }

    public String getCategory() {
        return category;
    }

    public void setCategory(String category) {
        this.category = category;
    }

    public Priority getPriority() {
        return priority;
    }

    public void setPriority(Priority priority) {
        this.priority = priority;
    }

    public boolean isDone() {
        return isDone;
    }

    public void setDone(boolean done) {
        isDone = done;
    }

    public long getArriveDate() {
        return arriveDate;
    }

    public void setArriveDate(long arriveDate) {
        this.arriveDate = arriveDate;
    }

    public DateTime getDateTime() {
        dateTime = new DateTime();

        if (arriveDate == 0)
            return dateTime;

        PersianPickerDate persianDate = new PersianDateImpl();
        persianDate.setDate(arriveDate);

        dateTime.setDate(persianDate);

        DateHelper dateHelper = new DateHelper(arriveDate);

        dateTime.setHour(dateHelper.getHour());
        dateTime.setMinute(dateHelper.getMinute());

        return dateTime;
    }

    public void setDateTime(DateTime dateTime) {
        this.dateTime = dateTime;
    }

    public String getPersianDate() {
        if (arriveDate == 0)
            return null;

        PersianPickerDate persianDate = new PersianDateImpl();
        persianDate.setDate(arriveDate);
        return persianDate.getPersianLongDate();
    }

    public String getClock() {
        DateHelper dateHelper = new DateHelper(arriveDate);
        return dateHelper.getClock();
    }

    @Override
    public int compareTo(Todo todo) {
        boolean isSame = id == todo.getId() &&
                Objects.equals(title, todo.getTitle()) &&
                Objects.equals(category, todo.category) &&
                categoryId == todo.getCategoryId() &&
                arriveDate == todo.getArriveDate() &&
                Objects.equals(priority, todo.priority) &&
                isDone == todo.isDone();

        if (isSame)
            return 0;

        return 1;
    }

    public enum Priority {
        LOW, NORMAL, HIGH
    }

}

File: app/src/main/java/model/Todo.kt
package model

import android.os.Parcelable
import androidx.annotation.Keep
import androidx.room.ColumnInfo
import androidx.room.Entity
import androidx.room.Ignore
import androidx.room.PrimaryKey
import ir.hamsaa.persiandatepicker.api.PersianPickerDate
import ir.hamsaa.persiandatepicker.date.PersianDateImpl
import kotlinx.parcelize.IgnoredOnParcel
import kotlinx.parcelize.Parcelize
import utils.DateHelper

@Keep
@Parcelize
@Entity(tableName = "todos")
data class Todo(
    @ColumnInfo(name = "id")
    @PrimaryKey(autoGenerate = true)
    var id: Int = 0,

    @ColumnInfo(name = "title")
    var title: String? = null,

    @ColumnInfo(name = "category_id")
    var categoryId: Int = 0,

    @ColumnInfo(name = "category")
    var category: String? = null,

    @ColumnInfo(name = "priority")
    var priority: Priority? = null,

    @ColumnInfo(name = "is_done")
    var isDone: Boolean = false,

    @ColumnInfo(name = "arrive_date", defaultValue = "0")
    var arriveDate: Long = 0,

    @ColumnInfo(name = "created_at", defaultValue = "0")
    var createdAt: Long = 0,

    @ColumnInfo(name = "updated_at", defaultValue = "0")
    var updatedAt: Long = 0,
) : Comparable<Todo>, Parcelable {

    @Ignore
    @IgnoredOnParcel
    var dateTime: DateTime? = null
        get() {
            field = DateTime()
            if (arriveDate == 0L) return field

            val persianDate: PersianPickerDate = PersianDateImpl()
            persianDate.setDate(arriveDate)

            field!!.date = persianDate

            val dateHelper = DateHelper(arriveDate)
            field!!.hour = dateHelper.getHour()
            field!!.minute = dateHelper.getMinute()

            return field
        }

    val clock: String
        get() {
            val dateHelper = DateHelper(arriveDate)
            return dateHelper.getClock()
        }

    override fun compareTo(other: Todo): Int {
        val isSame = id == other.id &&
                title == other.title &&
                category == other.category &&
                categoryId == other.categoryId &&
                arriveDate == other.arriveDate &&
                priority == other.priority &&
                isDone == other.isDone

        return if (isSame) 0 else 1
    }
}